lib/sshutils/sftp/parse.go

```diff
@@ -17,64 +17,119 @@ limitations under the License.
 package sftp
 
 import (
-	"regexp"
+	"strings"
 
 	"github.com/gravitational/trace"
 
 	"github.com/gravitational/teleport/lib/utils"
 )
 
-var reSFTP = regexp.MustCompile(
-	// optional username, note that outside group
-	// is a non-capturing as it includes @ signs we don't want
-	`(?:(?P<username>.+)@)?` +
-		// either some stuff in brackets - [ipv6]
-		// or some stuff without brackets and colons
-		`(?P<host>` +
-		// this says: [stuff in brackets that is not brackets] - loose definition of the IP address
-		`(?:\[[^@\[\]]+\])` +
-		// or
-		`|` +
-		// some stuff without brackets or colons to make sure the OR condition
-		// is not ambiguous
-		`(?:[^@\[\:\]]+)` +
-		`)` +
-		// after colon, there is a path that could consist technically of
-		// any char including empty which stands for the implicit home directory
-		`:(?P<path>.*)`,
-)
-
-// Destination is SCP destination to copy to or from
+// Destination is a remote SFTP destination to copy to or from.
 type Destination struct {
 	// Login is an optional login username
 	Login string
 	// Host is a host to copy to/from
-	Host utils.NetAddr
+	Host *utils.NetAddr
 	// Path is a path to copy to/from.
 	// An empty path name is valid, and it refers to the user's default directory (usually
 	// the user's home directory).
 	// See https://tools.ietf.org/html/draft-ietf-secsh-filexfer-09#page-14, 'File Names'
 	Path string
 }
 
-// ParseSCPDestination takes a string representing a remote resource for SFTP
-// to download/upload, like "user@host:/path/to/resource.txt" and parses it into
+// ParseDestination takes a string representing a remote resource for SFTP
+// to download/upload in the form "[user@]host:[path]" and parses it into
 // a structured form.
 //
 // See https://tools.ietf.org/html/draft-ietf-secsh-filexfer-09#page-14, 'File Names'
 // section about details on file names.
-func ParseDestination(s string) (*Destination, error) {
-	out := reSFTP.FindStringSubmatch(s)
-	if len(out) < 4 {
-		return nil, trace.BadParameter("failed to parse %q, try form user@host:/path", s)
+func ParseDestination(input string) (*Destination, error) {
+	firstColonIdx := strings.Index(input, ":")
+	if firstColonIdx == -1 {
+		return nil, trace.BadParameter("%q is missing a path, use form [user@]host:[path]", input)
 	}
-	addr, err := utils.ParseAddr(out[2])
-	if err != nil {
-		return nil, trace.Wrap(err)
+	hostStartIdx := strings.LastIndex(input[:firstColonIdx], "@")
```

Comment:
Files could contain `@` characters, it looks like the logic below would not parse correctly in this condition

Reply:
Paths can contain `@` characters, which is why I added a few test cases to test that: https://github.com/gravitational/teleport/blob/1df516900847d87c8371d92b813497c97ca9f452/lib/sshutils/sftp/parse_test.go#L121-L129

Reply:
It does look like you're handling the `@` correctly, my mistake.

Reply:
No worries, took me a bit to get this logic correct (and simple)